delivery_purchase/models/purchase_order.py

```diff
@@ -25,14 +26,71 @@ def _compute_carrier_id(self):
             carrier = item.partner_id.property_delivery_carrier_id or False
             item.carrier_id = carrier
 
-    @api.depends("order_line", "amount_total", "carrier_id")
+    @api.depends("order_line", "order_line.is_delivery", "amount_total", "carrier_id")
     def _compute_delivery_price(self):
         for item in self.filtered(lambda x: x.carrier_id):
-            item.delivery_price = item.carrier_id.purchase_rate_shipment(self)["price"]
+            delivery_lines = item.order_line.filtered(lambda x: x.is_delivery)
+            if delivery_lines:
+                item.delivery_price = sum(delivery_lines.mapped("price_unit"))
+            else:
+                item.delivery_price = item.carrier_id.purchase_rate_shipment(self)[
+                    "price"
+                ]
 
     @api.model
     def _prepare_picking(self):
         res = super()._prepare_picking()
         if self.carrier_id:
             res["carrier_id"] = self.carrier_id.id
         return res
+
+    @api.depends("order_line.is_delivery")
+    def _get_invoice_status(self):
+        super()._get_invoice_status()
+        for order in self:
+            if order.invoice_status in ["no", "invoiced"]:
+                continue
+            order_lines = order.order_line.filtered(
+                lambda x: not x.is_delivery and not x.display_type
+            )
+            if all(
+                line.product_id.invoice_policy == "delivery"
+                and line.invoice_status == "no"
+                for line in order_lines
+            ):
+                order.invoice_status = "no"
+
+    def _create_delivery_line(self, carrier, price_unit):
+        product = carrier.product_id
+        if product.description_purchase:
+            po_description = "%s: %s" % (carrier.name, product.description_purchase)
+        else:
+            po_description = carrier.name
+
+        taxes = product.supplier_taxes_id
+        if self.partner_id and self.fiscal_position_id:
```

Comment:
Why doing this manually instead of relying in standard? Call `onchange` for getting the proper values.

Reply:
I have changed the way of set the values to avoid all this with another approach, do you see it correct now?